### PR TITLE
Rewrite `Desugar.makePatDef` to fix typing issues and be more understandable

### LINKED ISSUE #24435: Compiler error when deconstructing tuple of literals
## Compiler version

3.3.7, 3.7.4, 2.13.17

## Minimized code

```scala
//> using scala 3.7.4

val (x: "x", y: "y") = (("x", "y"): ("x", "y"))
println(x)
```

## Output

```scala
[error] Found:    (run$_.this.$1$._1 : String)
[error] Required: ("x" : String)
[error] val (x: "x", y: "y") = (("x", "y"): ("x", "y"))
[error]            ^
[error] ./run.sc:6:20
[error] Found:    (run$_.this.$1$._2 : String)
[error] Required: ("y" : String)
[error] val (x: "x", y: "y") = (("x", "y"): ("x", "y"))
[error]                    ^
[warn] ./run.sc:6:25
[warn] pattern's type ("x" : String) is more specialized than the right hand side expression's type String
[warn] 
[warn] If the narrowing is intentional, this can be communicated by adding `: @unchecked` after the expression,
[warn] which may result in a MatchError at runtime.
[warn] This patch can be rewritten automatically under -rewrite -source 3.2-migration.
[warn] val (x: "x", y: "y") = (("x", "y"): ("x", "y"))
[warn]                         ^^^^^^^^^^^^^^^^^^^^^^
```

## Expectation
no error, no warn

Note that this code is compiling:
```scala
//> using scala 3.7.4

(("x", "y"): ("x", "y")) match
  case (x: "x", y: "y") => 
    println(x)
```

### PR DESCRIPTION
**Pull request by SolalPirelli**
Fixes #24435 

`makePatDef` was quite hard to understand and also wrong in edge cases related to types.

I rewrote it with (1) detailed comments of what the optimizations are, (2) support for more patterns such as `Typed(Tuple(_), _)`, and (3) a `printing` test that enshrines what we expect from these optimizations.

Because it's easier to understand, it enables further optimizations like #25010

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 1]
```diff
@@ -1467,166 +1467,151 @@ object desugar {
       sel
     end match
 
-  case class TuplePatternInfo(arity: Int, varNum: Int, wildcardNum: Int)
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala:1473]: Review in split diff, this is a rewrite

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -2300,14 +2281,8 @@ object desugar {
...
-          val pats = valeqs.map { case GenAlias(pat, _) => pat }
-          val rhss = valeqs.map { case GenAlias(_, rhs) => rhs }
-          val (defpats, ids) = pats.map(makeIdPat).unzip
-          val pdefs = valeqs.lazyZip(defpats).lazyZip(rhss).map { (valeq, defpat, rhs) =>
-            val mods = defpat match
-              case defTree: DefTree => defTree.mods
-              case _ => Modifiers()
-            makePatDef(valeq, mods, defpat, rhs)
+          val pdefs = valeqs.map { case valeq @ GenAlias(pat, rhs) =>
+            makePatDef(makeIdPat(pat)._1, rhs, valeq.span, givenPatternsAllowed = true)
```
Comment (SolalPirelli) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: No behavior changes in the calls, just cleaning up some weird "extract then immediately zip" patterns (perhaps left over from some refactorings?)

File: tests/neg/i4935.scala [hunk 3]
```diff
@@ -1,3 +1,3 @@
 object Foo {
-  val (A, B) = () // error // error
+  val (A, B) = () // error
```
Comment (SolalPirelli) [inline tests/neg/i4935.scala:2]: Not quite sure why there were 2 errors before but I think 1 is quite enough :)

File: tests/pos/tuple-of-literals-deconstruct.scala [hunk 4]
```diff
@@ -0,0 +1,5 @@
+object Test {
+  def test =
+    val (x: "x", y: "y") = (("x", "y"): ("x", "y"))
+    println(x)
+}
```
Comment (SolalPirelli) [inline tests/pos/tuple-of-literals-deconstruct.scala:5]: This is the original test from the issue this PR is fixing

File: tests/pos/tuple-unchecked-deconstruct.scala [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+class B extends A { }
+class C extends A { }
+
+object Test {
+  def foo(): (List[A], List[A]) = ???
+  def bar(bs: List[B], cs: List[C]): Unit = ()
+
+  def test =
+    val (x: List[B @unchecked], y: List[C @unchecked]) = foo(): @unchecked
+    bar(x, y)
```
Comment (SolalPirelli) [inline tests/pos/tuple-unchecked-deconstruct.scala:11]: This is a simplified version of code in `Typer.scala` that failed to compile in my initial attempt at fixing this

File: tests/printing/posttyper/patdef.check [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+             => Tuple2.apply[Int, Int](i, j)
+        }
+    val i: Int = $4$._1
+    val j: Int = $4$._2
+    def foo(): Tuple2[Int, Int] = ???
+    private[<empty>] val $6$: (Int, Int) =
+      foo():(Int, Int) @RuntimeChecked match
+        {
+          case $5$ @ Tuple2.unapply[Int, Int](_, _) => $5$:(Int, Int)
+        }
```
Comment (SolalPirelli) [inline tests/printing/posttyper/patdef.check:37]: It is somewhat annoying that this desugaring happens before typing, which means we can't tell that `foo()` is already a `Tuple2` here. We have some syntactic pattern matching to find basic cases, like the `if` example at the end of this test, but that ideally shouldn't be necessary...

File: tests/printing/posttyper/patdef.check [hunk 7]
```diff
@@ -0,0 +1,79 @@
+[[syntax trees at end of                 posttyper]] // tests/printing/posttyper/patdef.scala
```
Comment (SolalPirelli) [inline tests/printing/posttyper/patdef.check:1]: The results of running this test have no regressions compared to `main`, and the cases with typed tuples are better (variables `g, h, i, j, n, o, p` in source)

File: tests/pos/tuple-deconstruct-upcast.scala [hunk 8]
```diff
@@ -0,0 +1,8 @@
+object Test {
+  def consume(a: Array[AnyRef]): Unit = {}
+
+  def test =
+    val (x, y): (AnyRef, AnyRef) = ("a", "b")
+    val arr = Array(x)
+    consume(arr) // depends on `arr` being inferred as Array[AnyRef]
+}
```
Comment (SolalPirelli) [inline tests/pos/tuple-deconstruct-upcast.scala:8]: Simplified version of a ScalaJS unit test that failed to compile in one attempt at doing this

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 9]
```diff
@@ -1467,166 +1467,160 @@ object desugar {
...
-                  ValDef(named.name.asTermName, tpt, selector(n))
-                    .withMods(mods)
-                    .withSpan(named.span)
-                    .withAttachment(PatternVar, ())
-                )
-          flatTree(firstDef :: restDefs)
-      }
-  }
+        // Now that we have necessary info, define the actual RHS of the resulting assignment.
+        // We attach `ForArtifact` to tuples we create so our lowered code plays nice with "unused code" warnings later in the pipeline
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I have a branch for #25004 that touches this: the attachment is for when there is a Tuple. I'll provide feedback on the refactor, although I didn't try to understand the code yesterday.

Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I'm not 100% on what exactly this attachment does, which is why this comment is a bit vague, if you have more details I'd love to know. (I was wondering if this attachment was really necessary, there are indeed more warnings if I stop attaching it, but I didn't look at what those are and whether we could avoid them in some other way)

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Edit: actually, I see my previous refactor hadn't gone in yet. Finally approved last week. https://github.com/scala/scala3/pull/23448 I'll take a glance at how much work it makes for you.

Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Looks like the conflicts are minor in a couple calls to `makePatDef`, shouldn't be too hard (edit: done)

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 10]
```diff
@@ -1467,166 +1467,160 @@ object desugar {
...
-      // 2. Partial optimization:
-      //    If the rhs can be typed as a tuple and matched with correct arity, we can
-      //    return the tuple itself in the case if there are no more than one variable
-      //    in the pattern, or return the the value if there is only one variable.
-
-      val fullTupleOptimizable =
-        val isMatchingTuple: Tree => Boolean = {
-          case Tuple(es) => tuplePatternInfo.varNum == es.length && !hasNamedArg(es)
-          case _ => false
+  def makePatDef(pat: Tree, rhs: Tree, span: Span, givenPatternsAllowed: Boolean, overrideMods: Option[Modifiers] = None)(using Context): Tree =
```
Comment (som-snytt) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Instead of `Option[Modifiers]`, `original` to mean "mod-bearing tree" is more clear, it's the `DefTree`. (Its default could be `EmptyTree` or `pat` for fancy, or just require it.) (For `original`, it may not be clear "what properties are you deriving from that tree?") Probably OK then to provide `span: Span` explicitly.

Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: That's not what happens in main right now though, if you look at (old) lines 2288-2291, mods is from `pat` in these cases, whereas it's from `original` in the other calls.

Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I mean just have a Tree param that supplies the mod. Call it `treeBearingMods` if desired. In preference to an optional param. Rather than ask the caller to say `someTree.mods` or equivalent.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 9 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I was conservative in refactor because resistance to refactor. I'll have to blink the sleep out of my eyes first, but I do vaguely remember that intermediate values aren't terrible if they assist reading. (That is an Oderskyism.)

**Comment by SolalPirelli** [comment]
CI failed for the second time because `(Int, Int)` is printed as `Tuple2[Int, Int]` in the bootstrapped leg... looking into it

**Comment by som-snytt** [comment]
I opened a PR that implements my suggestion to reduce rather than augment the number of params to `makePatDef`.

https://github.com/scala/scala3/pull/25369

I don't know if it will pass yet, but my sense is that passing around booleans and spans is brittle or confusing; I restored `original` and hope it works.

I agree the refactor is easier to follow; the same code on Scala 2 is also convoluted. But I'm hoping that stretching my working memory will stave off dementia later.

The line comments are a lot of words. A lotta words.

I didn't delete words, but I restored the old comment, which is diagrammy.

They're talking about scalafmt, so since I'm using a 120-char terminal window, I took the liberty of keeping lines shorter.

I didn't try pushing to the branch, usually it says when pushes are welcome. Github interface changes frequently, so indicators and buttons move around.

**Comment by som-snytt** [comment]
My PR passes, but I don't have permission on this branch; feel free to fetch my changes.

Edit: my PR is already rebased.

The UI here doesn't show me the conflicts? I used to like github. I remember there was a minor conflict.

File: tests/neg/i4935.scala (hunk 3 above)
Comment (som-snytt) [reply tests/neg/i4935.scala:2]: Indeed, the arity error is a surface error, without knowing anything about A or B.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 2 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: I assume someone ported from Scala 2, which I'm afraid to look at, at the moment. Undoubtedly the weird zips looked beautiful at the time. On the other PR, I can see (for now, until I forget) how it got from the original to the edit, but it's a not a formulation I would have produced off the cuff.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 1 above)
Comment (som-snytt) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala:1473]: I tried that, but it just made the windows too narrow to read. 😄

**Review by som-snytt — APPROVED** [review]
The tests pass! I don't mind if my review PR is a follow-up, it's up to you. "It's only software."
